### PR TITLE
Added support for nested attributes

### PR DESCRIPTION
**Pull request by elij**
Good Afternoon, 

I've added the following code to use this library for creating VETH pairs -- it specifically allows nesting of attributes.

However I'm not sure if this is inline with code from this project.

Is it possible to get feedback?

Thanks

### CODE REVIEW

File: src/packet/route/route.rs [hunk 1]
```diff
@@ -283,6 +289,58 @@ impl<'a> ToPayload for RtAttrPacket<'a> {
...
+
+    fn payload_size(&self) -> usize {
+        match *self {
+            Nested::RtAttr(ref r) => util::align(r.packet_size()),
+            Nested::InfoAttr(ref i) => util::align(i.packet_size()),
+        }
+    }
+}
+
+impl<'a> ToPayload for &'a [&'a RtAttrPacket<'a>] {
```
Comment (bkchr) [inline src/packet/route/route.rs:331]: As we already implement impl<'a> ToPayload for &'a [&'a ToPayload, this implementation should already exist implicit.

File: src/packet/route/route.rs [hunk 2]
```diff
@@ -283,6 +289,58 @@ impl<'a> ToPayload for RtAttrPacket<'a> {
...
+    fn payload_add(&self, payload: &mut [u8]) {
+        payload[..self.len()].copy_from_slice(self)
+    }
+
+    fn payload_size(&self) -> usize {
+        self.len() + 1
+    }
+}
+
+impl<'a> ToPayload for Nested<'a> {
```
Comment (bkchr) [inline src/packet/route/route.rs:315]: I would accept `ToPayload` for `IfInfoPacket`. If you need this enum in your project, I would say you implement it in your project. I have done the same for a project of me.

File: src/packet/route/route.rs [hunk 3]
```diff
@@ -283,6 +289,58 @@ impl<'a> ToPayload for RtAttrPacket<'a> {
...
+    }
+}
+
+impl<'a> ToPayload for &'a [u8] {
+    fn payload_add(&self, payload: &mut [u8]) {
+        payload[..self.len()].copy_from_slice(self)
+    }
+
+    fn payload_size(&self) -> usize {
+        self.len() + 1
```
Comment (bkchr) [inline src/packet/route/route.rs:311]: Why plus one? I would expect that the length is encoded as is, don't know why it should have a NullByte at the end.

File: src/packet/route/route.rs [hunk 4]
```diff
@@ -283,6 +289,58 @@ impl<'a> ToPayload for RtAttrPacket<'a> {
     }
 }
 
+impl<'a> ToPayload for &'a [&'a Nested<'a>] {
```
Comment (bkchr) [inline src/packet/route/route.rs:292]: Can also be dropped, as you already implement `ToPayload` for `Nested`.

**Review by bkchr — CHANGES_REQUESTED** [review]
I added some comments, in general I'm unsure why we need this `Nested` enum in the project.

**Comment by elij** [comment]
Thanks I would also like to avoid adding the enum but outside of pnetlink I can't access the payload.

I need a way with a RtAttr to have a payload of an array mixed of IfInfoMsg and RtAttr (see use here: https://github.com/lypy/andover/blob/master/src/main.rs#L102)

VETH_INFO_PEER (RtAttr)
- IfInfo
- IFLA_IFNAME RtAttr)

Do you have any suggestions? I'll do the heavy lifting outside and limit the pull request to just the ifinfo payload.

**Comment by bkchr** [comment]
Did you tried:
```
let mut link_info_data : Vec<&ToPayload> = Vec::new();
link_info_data.push(&n1);
link_info_data.push(&n2);

let info_peer = RtAttrPacket::create_with_payload(VETH_INFO_PEER, &link_info_data[..]);
```

Look [here](https://github.com/superscale/worm/blob/152ce141cbae111a1a91574b9af51d077436311b/src/tunnel.rs#L56) to see what I'm doing to create a big package.

**Comment by elij** [comment]
Thank you very much :) 

I'll create a new pull request

**Pull request closed without merging**